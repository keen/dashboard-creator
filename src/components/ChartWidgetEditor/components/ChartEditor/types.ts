export enum ChartEditorError {
  WIDGET = 'WIDGET',
  CONFIGURATION = 'CONFIGURATION',
}
