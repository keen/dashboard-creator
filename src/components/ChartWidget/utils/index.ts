import mergeWidgetSettingsWithFont from './mergeWidgetSettingsWithFont';
import createDataviz from './createDataviz';

export { mergeWidgetSettingsWithFont, createDataviz };
