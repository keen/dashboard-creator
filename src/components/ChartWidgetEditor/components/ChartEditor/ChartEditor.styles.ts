import styled from 'styled-components';

export const Container = styled.div`
  background: #f1f5f8;
`;

export const VisualizationContainer = styled.div`
  padding: 10px;
`;

export const Footer = styled.div`
  display: flex;
  padding: 10px 25px;
`;

export const FooterAside = styled.div`
  display: flex;
  margin-left: auto;
`;
