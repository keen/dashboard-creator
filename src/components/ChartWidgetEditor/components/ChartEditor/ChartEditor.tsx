/* eslint-disable @typescript-eslint/no-empty-function */
import React, { FC, useState, useEffect, useRef, useContext } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import deepEqual from 'deep-equal';
import QueryCreator from '@keen.io/query-creator';
import { Button, Anchor, FadeLoader } from '@keen.io/ui-core';
import { Query } from '@keen.io/query';

import {
  Container,
  QueryCreatorContainer,
  VisualizationContainer,
  Cancel,
  Footer,
  FooterAside,
} from './ChartEditor.styles';

import {
  applyConfiguration,
  editorMounted,
  runQuery,
  setQuerySettings,
  setQueryChange,
  setVisualizationSettings,
  getChartEditor,
} from '../../../../modules/chartEditor';
import { getActiveDashboardTheme } from '../../../../modules/theme';

import WidgetVisualization from '../WidgetVisualization';
import { AppContext } from '../../../../contexts';

type Props = {
  onClose: () => void;
};

// @TODO: Support host as context argument

const ChartEditor: FC<Props> = ({ onClose }) => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const {
    modalContainer,
    project: { id, userKey, masterKey },
  } = useContext(AppContext);

  const {
    analysisResult,
    querySettings,
    visualization,
    isEditMode,
    isQueryPerforming,
  } = useSelector(getChartEditor);
  const baseTheme = useSelector(getActiveDashboardTheme);

  const queryRef = useRef(null);
  const [initialQuery, setInitialQuery] = useState(null);
  queryRef.current = initialQuery;

  useEffect(() => {
    dispatch(editorMounted());
  }, []);

  useEffect(() => {
    if (initialQuery) {
      const hasQueryChanged = !deepEqual(initialQuery, querySettings, {
        strict: true,
      });
      dispatch(setQueryChange(hasQueryChanged));
    }
  }, [querySettings, initialQuery]);

  return (
    <Container>
      <VisualizationContainer>
        <WidgetVisualization
          visualization={visualization}
          baseTheme={baseTheme}
          analysisResult={analysisResult}
          querySettings={querySettings}
          onChangeVisualization={({ type, chartSettings, widgetSettings }) =>
            dispatch(
              setVisualizationSettings(type, chartSettings, widgetSettings)
            )
          }
        />
      </VisualizationContainer>
      <QueryCreatorContainer>
        <QueryCreator
          projectId={id}
          readKey={userKey}
          masterKey={masterKey}
          modalContainer={modalContainer}
          onUpdateChartSettings={() => {}}
          onUpdateQuery={(query: Query) => {
            dispatch(setQuerySettings(query));
            if (!queryRef.current) {
              setInitialQuery(query);
            }
          }}
          host="staging-api.keen.io"
        />
      </QueryCreatorContainer>
      <Footer>
        <Button
          variant="success"
          isDisabled={isQueryPerforming}
          icon={isQueryPerforming && <FadeLoader />}
          onClick={() => dispatch(runQuery())}
        >
          {isQueryPerforming
            ? t('chart_widget_editor.run_query_loading')
            : t('chart_widget_editor.run_query')}
        </Button>
        <FooterAside>
          <Cancel>
            <Anchor onClick={onClose}>{t('chart_widget_editor.cancel')}</Anchor>
          </Cancel>
          <Button
            variant="secondary"
            isDisabled={isQueryPerforming}
            onClick={() => dispatch(applyConfiguration())}
          >
            {isEditMode
              ? t('chart_widget_editor.save')
              : t('chart_widget_editor.add_to_dashboard')}
          </Button>
        </FooterAside>
      </Footer>
    </Container>
  );
};
export default ChartEditor;
