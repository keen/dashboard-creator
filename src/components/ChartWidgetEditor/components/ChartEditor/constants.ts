import { ChartEditorError } from './types';

export const CHART_EDITOR_ERRORS: Record<ChartEditorError, string> = {
  [ChartEditorError.CONFIGURATION]: 'chart_widget_editor.configuration_error',
  [ChartEditorError.WIDGET]: 'chart_widget_editor.widget_error',
};
