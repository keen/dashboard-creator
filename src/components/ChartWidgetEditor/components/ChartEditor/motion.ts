export const notificationBarMotion = {
  initial: { top: '-100%' },
  animate: { top: 0 },
  exit: { top: '-100%' },
  transition: { duration: 0.3 },
};
