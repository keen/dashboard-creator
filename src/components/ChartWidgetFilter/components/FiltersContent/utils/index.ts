import capitalize from './capitalize';
import getPropertyName from './getPropertyName';

export { capitalize, getPropertyName };
