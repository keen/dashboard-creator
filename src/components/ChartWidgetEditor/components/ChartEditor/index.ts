import ChartEditor from './ChartEditor';

export default ChartEditor;
