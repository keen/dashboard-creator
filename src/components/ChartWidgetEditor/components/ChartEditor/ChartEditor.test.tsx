/* eslint-disable @typescript-eslint/naming-convention */
import React from 'react';
import {
  render as rtlRender,
  fireEvent,
  waitFor,
  cleanup,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { PubSub } from '@keen.io/pubsub';

import ChartEditor from './ChartEditor';

import { AppContext } from '../../../../contexts';
import { createBodyElementById } from '../../../../utils/test/createBodyElementById';

import {
  chartEditorActions,
  chartEditorInitialState,
  EditorSection,
} from '../../../../modules/chartEditor';
import { createDashboardSettings } from '../../../../modules/dashboards/utils';

const render = (storeState: any = {}, overProps: any = {}) => {
  const props = {
    onClose: jest.fn(),
    ...overProps,
  };

  const activeDashboardId = '@dashboard-id';

  const state = {
    app: { activeDashboardId },
    theme: {
      dashboards: {
        [activeDashboardId]: {
          theme: {
            gridX: {
              enabled: true,
            },
            gridY: {
              enabled: false,
            },
            funnel: {
              header: {
                badge: {
                  enabled: true,
                },
              },
            },
          },
          settings: createDashboardSettings(),
        },
      },
    },
    chartEditor: chartEditorInitialState,
    timezone: {
      defaultTimezoneForQuery: 'Africa/Nairobi',
      timezoneSelectionDisabled: false,
    },
    ...storeState,
  };

  const mockStore = configureStore([]);
  const store = mockStore({ ...state });

  const appContext = {
    notificationPubSub: new PubSub(),
    analyticsApiUrl: '@keen-api-url',
    timezonesApiUrl: '@timezones-api-url',
    modalContainer: '#modal-root',
    project: {
      id: '@project-id',
      userKey: '@user-key',
      masterKey: '@master-key',
    },
    createSharedDashboardUrl: () => 'url',
  };

  const wrapper = rtlRender(
    <Provider store={store}>
      <AppContext.Provider value={appContext}>
        <ChartEditor {...props} />
      </AppContext.Provider>
    </Provider>
  );

  return {
    store,
    appContext,
    props,
    wrapper,
  };
};

jest.useFakeTimers();

afterEach(() => {
  cleanup();
});

beforeEach(() => {
  createBodyElementById('modal-root');
});

test('allows user to run query', () => {
  const {
    wrapper: { getAllByText },
    store,
  } = render();

  store.clearActions();

  const button = getAllByText('chart_widget_editor.run_query');
  fireEvent.click(button[0]);

  expect(store.getActions()).toMatchInlineSnapshot(`
    Array [
      Object {
        "payload": undefined,
        "type": "chartEditor/runQuery",
      },
    ]
  `);
});

test('triggers action after editor is mounted', async () => {
  const { store } = render();

  await waitFor(() => {
    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        {
          type: chartEditorActions.editorMounted.type,
        },
      ])
    );
  });
});

test('do not allows user to apply incomplete chart settings', () => {
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: null,
      visualization: {
        type: 'metric',
        chartSettings: {},
        widgetSettings: {},
      },
    },
  };

  const {
    wrapper: { getByText },
    store,
  } = render(storeState);

  store.clearActions();

  const button = getByText('chart_widget_editor.add_to_dashboard');
  fireEvent.click(button);

  expect(store.getActions()).toEqual([]);
});

test('allows user to apply chart editor configuration', () => {
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: {
        query: {
          analysis_type: 'count',
        },
        results: 100,
      },
      visualization: {
        type: 'metric',
        chartSettings: {},
        widgetSettings: {},
      },
    },
  };

  const {
    wrapper: { getByText },
    store,
  } = render(storeState);

  store.clearActions();

  const button = getByText('chart_widget_editor.add_to_dashboard');
  fireEvent.click(button);

  expect(store.getActions()).toMatchInlineSnapshot(`
    Array [
      Object {
        "payload": undefined,
        "type": "chartEditor/applyConfiguration",
      },
    ]
  `);
});

test('renders visualization settings error', () => {
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: {
        query: {
          analysis_type: 'count',
        },
        results: 100,
      },
      visualization: {
        type: 'line',
        chartSettings: {},
        widgetSettings: {},
      },
    },
  };

  const {
    wrapper: { getByText },
  } = render(storeState);

  const button = getByText('chart_widget_editor.add_to_dashboard');
  fireEvent.click(button);

  expect(getByText('chart_widget_editor.widget_error')).toBeInTheDocument();

  jest.clearAllTimers();
});

test('calls "onClose" event handler', () => {
  const {
    props,
    wrapper: { getByText },
  } = render();

  const cancelButton = getByText('chart_widget_editor.cancel');
  fireEvent.click(cancelButton);

  expect(props.onClose).toHaveBeenCalled();
});

test('clicking "close" icon calls "onClose" handler', () => {
  const {
    props,
    wrapper: { getByTestId },
  } = render();

  const closeIcon = getByTestId('close-handler');
  fireEvent.click(closeIcon);

  expect(props.onClose).toHaveBeenCalled();
});

test('shows placeholder with run query button', () => {
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: null,
    },
  };
  const {
    wrapper: { getAllByText },
  } = render(storeState);

  expect(getAllByText('chart_widget_editor.run_query').length).toEqual(2);
});

test('does not allow user to apply chart editor configuration when query has error', () => {
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: {
        query: {
          analysis_type: 'count',
        },
        results: 100,
      },
      visualization: {
        type: 'metric',
        chartSettings: {},
        widgetSettings: {},
      },
      queryError: 'There is error in the query',
    },
  };

  const {
    wrapper: { getByText },
    store,
  } = render(storeState);

  store.clearActions();

  const button = getByText('chart_widget_editor.add_to_dashboard');
  fireEvent.click(button);
  expect(store.getActions()).toEqual([]);
});

test('does not allow user to apply chart editor configuration when query is dirty', () => {
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: {
        query: {
          analysis_type: 'count',
        },
        results: 100,
      },
      visualization: {
        type: 'metric',
        chartSettings: {},
        widgetSettings: {},
      },
      isDirtyQuery: true,
    },
  };

  const {
    wrapper: { getByText },
    store,
  } = render(storeState);

  store.clearActions();

  const button = getByText('chart_widget_editor.add_to_dashboard');
  fireEvent.click(button);
  expect(store.getActions()).toEqual([]);
});

test('allows user to set chart title', async () => {
  const chartTitle = 'CHART_TITLE';
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: {
        query: {
          analysis_type: 'count',
        },
        results: 100,
      },
      visualization: {
        type: 'area',
        visualizationType: 'area',
        chartSettings: {
          curve: 'linear',
          stackMode: 'normal',
          groupMode: 'grouped',
        },
        widgetSettings: {},
      },
      editorSection: EditorSection.SETTINGS,
    },
  };
  const {
    wrapper: { getByPlaceholderText },
    store,
  } = render(storeState);

  store.clearActions();
  const titleInput = getByPlaceholderText(
    'widget_customization_heading_settings.title_placeholder'
  );
  fireEvent.change(titleInput, { target: { value: chartTitle } });

  expect(store.getActions()).toMatchSnapshot();
});

test('allows user to set chart subtitle', async () => {
  const chartSubtitle = 'CHART_SUBTITLE';
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: {
        query: {
          analysis_type: 'count',
        },
        results: 100,
      },
      visualization: {
        type: 'area',
        visualizationType: 'area',
        chartSettings: {
          curve: 'linear',
          stackMode: 'normal',
          groupMode: 'grouped',
        },
        widgetSettings: {},
      },
      editorSection: EditorSection.SETTINGS,
    },
  };
  const {
    wrapper: { getByPlaceholderText },
    store,
  } = render(storeState);

  store.clearActions();
  const titleInput = getByPlaceholderText(
    'widget_customization_heading_settings.subtitle_placeholder'
  );
  fireEvent.change(titleInput, { target: { value: chartSubtitle } });

  expect(store.getActions()).toMatchSnapshot();
});

test('allows user use chart name from saved query', async () => {
  const storeState = {
    chartEditor: {
      ...chartEditorInitialState,
      analysisResult: {
        query_name: 'QUERY_NAME',
        metadata: {
          display_name: 'QUERY NAME',
        },
        query: {
          analysis_type: 'count',
        },
        results: 100,
      },
      visualization: {
        type: 'area',
        visualizationType: 'area',
        chartSettings: {
          curve: 'linear',
          stackMode: 'normal',
          groupMode: 'grouped',
        },
        widgetSettings: {},
      },
      editorSection: EditorSection.SETTINGS,
    },
  };
  const {
    wrapper: { getByTestId },
  } = render(storeState);

  const inheritQueryNameInfo = getByTestId('inherit-query-name');
  expect(inheritQueryNameInfo).toBeInTheDocument();
});
