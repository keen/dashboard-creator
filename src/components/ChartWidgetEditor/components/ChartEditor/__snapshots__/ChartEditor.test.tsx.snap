// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows user to set chart subtitle 1`] = `
Array [
  Object {
    "payload": true,
    "type": "chartEditor/setQueryChange",
  },
  Object {
    "payload": Object {
      "card": Object {
        "enabled": true,
      },
      "legend": Object {
        "alignment": "left",
        "enabled": true,
        "layout": "horizontal",
        "position": "top",
        "typography": Object {
          "fontColor": "#1D2729",
          "fontFamily": "Gangster Grotesk, Arial, Helvetica, sans-serif",
          "fontSize": 10,
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      },
      "subtitle": Object {
        "content": "CHART_SUBTITLE",
        "typography": Object {
          "fontColor": "#1D2729",
          "fontFamily": "Gangster Grotesk, Arial, Helvetica, sans-serif",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      },
      "title": Object {
        "content": "",
        "typography": Object {
          "fontColor": "#27566D",
          "fontFamily": "Gangster Grotesk, Arial, Helvetica, sans-serif",
          "fontSize": 20,
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      },
    },
    "type": "chartEditor/updateWidgetSettings",
  },
]
`;

exports[`allows user to set chart title 1`] = `
Array [
  Object {
    "payload": true,
    "type": "chartEditor/setQueryChange",
  },
  Object {
    "payload": Object {
      "card": Object {
        "enabled": true,
      },
      "legend": Object {
        "alignment": "left",
        "enabled": true,
        "layout": "horizontal",
        "position": "top",
        "typography": Object {
          "fontColor": "#1D2729",
          "fontFamily": "Gangster Grotesk, Arial, Helvetica, sans-serif",
          "fontSize": 10,
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      },
      "subtitle": Object {
        "content": "",
        "typography": Object {
          "fontColor": "#1D2729",
          "fontFamily": "Gangster Grotesk, Arial, Helvetica, sans-serif",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      },
      "title": Object {
        "content": "CHART_TITLE",
        "typography": Object {
          "fontColor": "#27566D",
          "fontFamily": "Gangster Grotesk, Arial, Helvetica, sans-serif",
          "fontSize": 20,
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      },
    },
    "type": "chartEditor/updateWidgetSettings",
  },
]
`;
